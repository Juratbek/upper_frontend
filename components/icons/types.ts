export interface IIconProps {
  color?: string;
  variant?: 'outlined' | 'fulfilled';
  width?: number;
  height?: number;
  opacity?: number;
}
