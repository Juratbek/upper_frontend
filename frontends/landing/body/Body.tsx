import { ApiErrorBoundary } from 'components';
import { Spinner } from 'components/lib';
import {
  LABEL_ID_PARAM,
  Labels,
  LastPublished,
  LoadMoreButton,
  NoArticle,
  PublishedArticle,
  TopLabel,
} from 'components/molecules';
import { Advertisement, Sidebar } from 'components/organisms';
import { useDevice, useUrlParams } from 'hooks';
import { usePublishedArticlesList } from 'store/clients/published-article';
import { addAmazonBucketUrl } from 'utils/published-article';
import { PopularLabels } from 'variables';

import classes from './Body.module.scss';

export const Body = (): JSX.Element => {
  const { setParam, getParam } = useUrlParams();
  const { isDesktop } = useDevice();
  const activeLabel = (getParam(LABEL_ID_PARAM) ?? TopLabel) as string;
  const {
    list: articles,
    fetchNextPage,
    ...articlesListRes
  } = usePublishedArticlesList(activeLabel);

  const labelSelectHandler = (label: string): void => {
    setParam(LABEL_ID_PARAM, label);
  };

  return (
    <div className={classes.root}>
      <section className={classes.container}>
        <Labels
          labels={[LastPublished, TopLabel, ...PopularLabels]}
          activeLabel={activeLabel}
          onSelect={labelSelectHandler}
        />
        {!isDesktop && <Advertisement />}
        <ApiErrorBoundary
          res={articlesListRes}
          fallback={
            <div style={{ height: '15rem' }} className='content-center'>
              <Spinner />
            </div>
          }
        >
          {articles.map(addAmazonBucketUrl).map((article) => (
            <PublishedArticle article={article} key={article.id} />
          ))}
          {articles.length === 0 && <NoArticle label={activeLabel} />}
          {articles.length !== 0 && articlesListRes.hasNextPage && (
            <LoadMoreButton loading={articlesListRes.isLoading} onClick={fetchNextPage} />
          )}
        </ApiErrorBoundary>
      </section>
      <div className={classes.sidebar}>
        <Sidebar />
      </div>
    </div>
  );
};
