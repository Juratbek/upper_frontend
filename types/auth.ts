import { Profile } from 'next-auth';
import { BuiltInProviderType } from 'next-auth/providers';
import { ClientSafeProvider, LiteralUnion } from 'next-auth/react';
import { AUTH_PROVIDER_TYPES } from 'variables';

type TGitHubUserType = 'User';

export type TAuthProviderProp = Profile & Record<string, unknown>;

export type TAuthProviderTypes =
  | typeof AUTH_PROVIDER_TYPES.google
  | typeof AUTH_PROVIDER_TYPES.facebook
  | typeof AUTH_PROVIDER_TYPES.github;

export type TAuthProviders = Record<LiteralUnion<BuiltInProviderType, string>, ClientSafeProvider>;

export interface IAuthProvider {
  [name: string]: (profile: TAuthProviderProp) => boolean;
}

export interface IGitHubUser {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: TGitHubUserType;
  site_admin: boolean;
  name?: string;
  company?: string;
  blog: string;
  location?: string;
  email?: string;
  hireable?: boolean;
  bio?: string;
  twitter_username?: string;
  public_repos: number;
  public_gists: number;
  followers: number;
  following: number;
  created_at: string;
  updated_at: string;
}

export interface IGoogleUser {
  iss: string;
  azp: string;
  aud: string;
  sub: string;
  email: string;
  email_verified: boolean;
  at_hash: string;
  name: string;
  picture: string;
  given_name: string;
  family_name: string;
  locale: string;
  iat: number;
  exp: number;
}

export interface IFacebookUser {
  id: string;
  name: string;
  email: string;
  picture: {
    data: {
      height: number;
      is_silhouette: boolean;
      url: string;
      width: number;
    };
  };
}

export interface ITelegramUser {
  auth_date: number;
  first_name: string;
  last_name?: string;
  photo_url: string;
  hash: string;
  id: number;
  username: string;
}
