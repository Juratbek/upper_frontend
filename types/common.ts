import { FC } from 'react';
import { ACTION_TYPES, ICON_TYPES } from 'variables';

export interface ITabHeader {
  name: string;
  id: string;
  active?: boolean;
}

export interface ITabBody {
  [tabId: string]: FC;
}

export type TAction = typeof ACTION_TYPES.delete | typeof ACTION_TYPES.markAsRead;

export type TIcon =
  | typeof ICON_TYPES.delete
  | typeof ICON_TYPES.save
  | typeof ICON_TYPES.home
  | typeof ICON_TYPES.user
  | typeof ICON_TYPES.menuList
  | typeof ICON_TYPES.notification
  | typeof ICON_TYPES.pen
  | typeof ICON_TYPES.comment
  | typeof ICON_TYPES.like
  | typeof ICON_TYPES.dislike
  | typeof ICON_TYPES.share
  | typeof ICON_TYPES.next
  | typeof ICON_TYPES.prev;

export interface IIcon {
  [name: TIcon]: FC<{ color?: string }>;
}

export type TClassName = string | undefined;
