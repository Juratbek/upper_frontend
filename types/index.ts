export * from './common';
export * from './article';
export * from './blog';
export * from './label';
export * from './notification';
export * from './auth';
