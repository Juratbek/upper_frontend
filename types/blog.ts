import { ILink } from './common';

export interface IBlogSmall {
  id: number;
  imgUrl: string;
  name: string;
}
export interface IBlogMedium extends IBlogSmall {
  bio?: string;
  followersCount?: number;
  articlesCount?: number;
}

export interface IBlog extends IBlogMedium {
  links: ILink[];
  createdDate: string;
  isFollowed?: boolean;
}
