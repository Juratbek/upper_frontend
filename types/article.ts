import { OutputData } from '@editorjs/editorjs';

import { IBlog } from './blog';
import { ILabel } from './label';

export interface IArticleResult {
  id: number;
  title: string;
  content: string;
  imgUrl?: string;
  author?: IBlog;
  labels: ILabel[];
  publishedDate: Date;
  updatedDate?: Date;
  viewCount?: number;
  likes?: number;
}

export interface IArticle {
  id: number;
  title: string;
  content: OutputData;
  imgUrl?: string;
  author?: IBlog;
  labels: ILabel[];
  publishedDate: string;
  updatedDate?: string;
  viewCount?: number;
  likes?: number;
  prevArticleId?: number;
  nextArticleId?: number;
}
