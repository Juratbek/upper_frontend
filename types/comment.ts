import { IBlogSmall } from './blog';

export interface IComment {
  id: number;
  date: string;
  text: string;
  author: IBlogSmall;
}
