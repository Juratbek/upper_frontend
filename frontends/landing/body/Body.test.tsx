import { render } from '@testing-library/react';
import { QueryWrapper, ReduxWrapper } from 'tests';
import { IPublishedArticleItem } from 'types';
import { vi } from 'vitest';

import { Body } from './Body';

const mockAuthor: IPublishedArticleItem['author'] = {
  id: 12,
  imgUrl: 'https://image.upper.uz/12',
  name: "Jur'atbek",
};

const mockArticles: IPublishedArticleItem[] = [
  {
    id: 1,
    imgUrl: 'https://image.upper.uz/1',
    title: 'Published article',
    author: mockAuthor,
  },
  {
    id: 2,
    imgUrl: 'https://image.upper.uz/2',
    title: 'Second published article',
    author: mockAuthor,
  },
  {
    id: 3,
    imgUrl: '',
    title: 'Third published article',
    author: mockAuthor,
  },
];

const mocks = vi.hoisted(() => ({
  usePublishedArticlesList: vi.fn().mockReturnValue({
    fetchNextPage: vi.fn(),
    list: [],
    hasNextPage: false,
    isLoading: true,
    isPending: true,
  }),
}));

vi.mock('store/clients/published-article', () => ({
  usePublishedArticlesList: mocks.usePublishedArticlesList,
}));

const Component = (
  <ReduxWrapper>
    <QueryWrapper>
      <Body />
    </QueryWrapper>
  </ReduxWrapper>
);

describe('Landing page Body', () => {
  it('snapshot', () => {
    render(Component);
    expect(document.body).toMatchSnapshot();
  });

  it('snapshot with articles returned by backend', () => {
    mocks.usePublishedArticlesList.mockReturnValue({
      list: mockArticles,
      fetchNextPage: vi.fn(),
      isLoading: false,
      isPending: false,
      isSuccess: true,
      hasNextPage: true,
    });

    render(Component);
    expect(document.body).toMatchSnapshot();
  });

  it('snapshot with empty articles returned by backend', () => {
    mocks.usePublishedArticlesList.mockReturnValue({
      list: [],
      fetchNextPage: vi.fn(),
      hasNextPage: false,
      isLoading: false,
      isPending: false,
      isSuccess: true,
    });

    render(Component);
    expect(document.body).toMatchSnapshot();
  });
});
