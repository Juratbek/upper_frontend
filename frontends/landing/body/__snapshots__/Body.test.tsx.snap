// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Landing page Body > snapshot 1`] = `
<body>
  <div>
    <div
      class="root"
    >
      <section
        class="container"
      >
        <div
          class="root"
          style="padding-top: 2rem; margin-top: -2rem;"
        >
          <div
            class="container"
          >
            <div
              class="labels-container"
              id="labels"
            >
              <div
                class="lebels-list"
              >
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Eng so'nggi
                </button>
                <button
                  class="button primary"
                  color="primary"
                >
                  Top
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Java
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  JavaScript
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  TypeScript
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  HTML
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  SQL
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Python
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C++
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Go
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C#
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  PHP
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Swift
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Kotlin
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Rust
                </button>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="content-center loading"
          >
            <div
              class="spinner spinner--primary"
              role="spinner"
            />
          </div>
        </div>
      </section>
      <div
        class="sidebar"
        style="top: 2rem;"
      >
        <div
          class="add-container"
        >
          <h4
            class="sale"
          >
            Chegirma
          </h4>
          <h3
            class="percent"
          >
            15%
          </h3>
          <p
            class="text"
          >
            UPPER promokodi bilan Frontint.uz kurslariga chegirma oling
          </p>
        </div>
        <a
          href="https://frontint.uz/"
          rel="noreferrer"
          target="_blank"
        >
          <button
            class="button mt-1 w-100 btn"
            style="opacity: 1;"
          >
            Chegirma olish
          </button>
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`Landing page Body > snapshot with articles returned by backend 1`] = `
<body>
  <div>
    <div
      class="root"
    >
      <section
        class="container"
      >
        <div
          class="root"
          style="padding-top: 2rem; margin-top: -2rem;"
        >
          <div
            class="container"
          >
            <div
              class="labels-container"
              id="labels"
            >
              <div
                class="lebels-list"
              >
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Eng so'nggi
                </button>
                <button
                  class="button primary"
                  color="primary"
                >
                  Top
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Java
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  JavaScript
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  TypeScript
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  HTML
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  SQL
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Python
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C++
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Go
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C#
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  PHP
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Swift
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Kotlin
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Rust
                </button>
              </div>
            </div>
          </div>
        </div>
        <div>
          <a
            class="article link"
            href="/web/articles/1"
          >
            <div
              class="article-img article-img article-img--full"
            >
              <img
                class="img"
                src="https://image.upper.uz/1"
              />
            </div>
            <h2
              class="article-title"
            >
              Published article
            </h2>
            <div
              class="article-footer"
            >
              <a
                class="link link"
                href="/web/blogs/12"
              >
                <div
                  class="blog"
                >
                  <div
                    class="avatar avatar--micro avatar"
                    style="border-color: rgba(245, 245, 245, 0.1);"
                  >
                    <img
                      alt="UPPER"
                      data-nimg="fill"
                      decoding="async"
                      loading="lazy"
                      sizes="100vw"
                      src="http://localhost:3000/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75"
                      srcset="/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75 3840w"
                      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
                    />
                  </div>
                  <h5
                    class="name"
                  >
                    Jur'atbek
                  </h5>
                </div>
              </a>
            </div>
          </a>
          <a
            class="article link"
            href="/web/articles/2"
          >
            <div
              class="article-img article-img article-img--full"
            >
              <img
                class="img"
                src="https://image.upper.uz/2"
              />
            </div>
            <h2
              class="article-title"
            >
              Second published article
            </h2>
            <div
              class="article-footer"
            >
              <a
                class="link link"
                href="/web/blogs/12"
              >
                <div
                  class="blog"
                >
                  <div
                    class="avatar avatar--micro avatar"
                    style="border-color: rgba(245, 245, 245, 0.1);"
                  >
                    <img
                      alt="UPPER"
                      data-nimg="fill"
                      decoding="async"
                      loading="lazy"
                      sizes="100vw"
                      src="http://localhost:3000/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75"
                      srcset="/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75 3840w"
                      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
                    />
                  </div>
                  <h5
                    class="name"
                  >
                    Jur'atbek
                  </h5>
                </div>
              </a>
            </div>
          </a>
          <a
            class="article link"
            href="/web/articles/3"
          >
            <h2
              class="article-title"
            >
              Third published article
            </h2>
            <div
              class="article-footer"
            >
              <a
                class="link link"
                href="/web/blogs/12"
              >
                <div
                  class="blog"
                >
                  <div
                    class="avatar avatar--micro avatar"
                    style="border-color: rgba(245, 245, 245, 0.1);"
                  >
                    <img
                      alt="UPPER"
                      data-nimg="fill"
                      decoding="async"
                      loading="lazy"
                      sizes="100vw"
                      src="http://localhost:3000/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75"
                      srcset="/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75 3840w"
                      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
                    />
                  </div>
                  <h5
                    class="name"
                  >
                    Jur'atbek
                  </h5>
                </div>
              </a>
            </div>
          </a>
          <button
            class="button"
            color="tertiary"
            style="opacity: 1; font-size: 20px; width: 100%;"
          >
            Yana yuklash
          </button>
        </div>
      </section>
      <div
        class="sidebar"
        style="top: 2rem;"
      >
        <div
          class="add-container"
        >
          <h4
            class="sale"
          >
            Chegirma
          </h4>
          <h3
            class="percent"
          >
            15%
          </h3>
          <p
            class="text"
          >
            UPPER promokodi bilan Frontint.uz kurslariga chegirma oling
          </p>
        </div>
        <a
          href="https://frontint.uz/"
          rel="noreferrer"
          target="_blank"
        >
          <button
            class="button mt-1 w-100 btn"
            style="opacity: 1;"
          >
            Chegirma olish
          </button>
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`Landing page Body > snapshot with empty articles returned by backend 1`] = `
<body>
  <div>
    <div
      class="root"
    >
      <section
        class="container"
      >
        <div
          class="root"
          style="padding-top: 2rem; margin-top: -2rem;"
        >
          <div
            class="container"
          >
            <div
              class="labels-container"
              id="labels"
            >
              <div
                class="lebels-list"
              >
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Eng so'nggi
                </button>
                <button
                  class="button primary"
                  color="primary"
                >
                  Top
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Java
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  JavaScript
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  TypeScript
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  HTML
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  SQL
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Python
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C++
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Go
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  C#
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  PHP
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Swift
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Kotlin
                </button>
                <button
                  class="button outlined"
                  color="outlined"
                >
                  Rust
                </button>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div>
            <h3>
              Top
               uchun maqolalar topilmadi
            </h3>
            <p>
              Top
               mavzusida maqolalar yozing va bilim ulashing
            </p>
          </div>
        </div>
      </section>
      <div
        class="sidebar"
        style="top: 2rem;"
      >
        <div
          class="add-container"
        >
          <h4
            class="sale"
          >
            Chegirma
          </h4>
          <h3
            class="percent"
          >
            15%
          </h3>
          <p
            class="text"
          >
            UPPER promokodi bilan Frontint.uz kurslariga chegirma oling
          </p>
        </div>
        <a
          href="https://frontint.uz/"
          rel="noreferrer"
          target="_blank"
        >
          <button
            class="button mt-1 w-100 btn"
            style="opacity: 1;"
          >
            Chegirma olish
          </button>
        </a>
      </div>
    </div>
  </div>
</body>
`;
